concurrency_test.go

```diff
@@ -0,0 +1,205 @@
+		go func() {
+			defer wg.Done()
+			statsigUser := User{
+				UserID:            fmt.Sprintf("statsig_u_%d", g),
+				Email:             "u@statsig.com",
+				PrivateAttributes: map[string]interface{}{"private": "shh"},
+				Custom:            map[string]interface{}{"key": "value"},
+				CustomIDs:         map[string]string{"randomID": "123456"},
+			}
+			user := User{
+				UserID: "regular_user_id",
+				Email:  "u@gmail.com",
+			}
+			for i := 0; i < loops; i++ {
+				LogEvent(Event{EventName: "test_event", User: user})
+				if CheckGate(statsigUser, "on_for_id_list") || !CheckGate(user, "on_for_id_list") {
+					t.Error("statsigUser should fail and regular user should pass for on_for_id_list gate")
+				}
+
+				if !CheckGate(statsigUser, "on_for_statsig_email") || CheckGate(user, "on_for_statsig_email") {
+					t.Error("statsig user should pass statsig email gate and regular user should fail")
+				}
+
+				exp := GetExperiment(statsigUser, "sample_experiment")
+				if !exp.GetBool("layer_param", false) {
+					t.Error("sample_experiment layer_param not correct")
+				}
+				config := GetConfig(statsigUser, "test_config")
+				if config.GetNumber("number", 420) != 7 {
+					t.Error("test_config number not correct")
+				}
+				layer := GetLayer(statsigUser, "a_layer")
+				if !layer.GetBool("layer_param", false) {
+					t.Error("sample_experiment layer_param not correct")
+				}
+				LogEvent(Event{EventName: "test_event_2", User: statsigUser})
+			}
+		}()
+	}
+	wg.Wait()
+
+	// 10 go routines x 10 loops each x 9 events (2 log event + 7 exposure events) = 900
+	if len(instance.logger.events) != 900 {
+		t.Error("Some events were dropped when calling APIs concurrently")
+	}
+
+	Shutdown()
+	if len(flushedEvents) != 900 {
+		t.Error("Not all events were flushed at shutdown")
+	}
+}
+
+func TestUpdatingRulesAndFetchingValuesConcurrently(t *testing.T) {
+	configSyncCount := 0
+	idlistSyncCount := 0
+	idListContent := "+7/rrkvF6\n" // hashed value for "regular_user_id", which is used for a user below
+	idListSize := len(idListContent)
+
+	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		defer req.Body.Close()
+		res.WriteHeader(http.StatusOK)
+		if strings.Contains(req.URL.Path, "download_config_specs") {
+			configSyncCount++
+			var in *downloadConfigsInput
+			bytes, _ := ioutil.ReadFile("download_config_specs.json")
+			json.NewDecoder(req.Body).Decode(&in)
+			res.Write(bytes)
+		} else if strings.Contains(req.URL.Path, "get_id_lists") {
+			idlistSyncCount++
+			baseURL := "http://" + req.Host
+			r := map[string]idList{
+				"list_1": {Name: "list_1", Size: int64(idListSize), URL: baseURL + "/list_1", CreationTime: 3, FileID: "file_id_1"},
+			}
+
+			v, _ := json.Marshal(r)
+			res.Write(v)
+		} else if strings.Contains(req.URL.Path, "list_1") {
+			res.Write([]byte(idListContent))
+			idListContent = fmt.Sprintf("+%d\n-%d\n", 1, 1)
+			idListSize += len(idListContent)
+		}
```

Comment:
hmm does it matter that we are writing +1/-1 to the end repeatedly, and only reading the initial content?  What is the exact race condition we are trying to test in id lists here?

Reply:
didn't mean to put 1 here always (updated), but either way, the idea is that the map for the ids get updated constantly, and the checkGate calls below are checking the same map constantly, and there is no crash or incorrect values